### DISCUSSION TITLE
Nx 2025 Roadmap

### DISCUSSION THREAD

**Discussion by jaysoo** [nrwl/nx#28731] [General]
# Nx 2025 Roadmap

As an open source project, Nx's commits and PRs are fully visible on GitHub. To provide more insight into the project's vision and direction, this roadmap highlights the key areas that the Nx team will focus on throughout 2025.

Note that this is a high-level roadmap, with features not tied to any specific future version of Nx. We may also update this roadmap as needed to make any course corrections.

## Nx Core in Rust

[Back in 2023](https://nx.dev/blog/nx-highlights-of-2023#typescript-for-extensibility-rust-for-speed), we started rewriting performance-critical parts of Nx core in Rust, such as caching and hashing. The move to Rust already shown speed benefits when running tasks in Nx, but we can go even further.

In 2025, we will bring the Rust story to its logical conclusion and port the rest of Nx core (i.e. the`nx` package itself, including the CLI) into Rust. This means that if you don't use any Nx plugins, then Nx itself will no longer require Node. Node is a fantastic platform, however there are fixed costs of invoking Node and importing modules that cannot be optimized further. 

This move also enables easier adoption of Nx into non-JavaScript monorepos. And I also want to reiterate that we will continue to use TypeScript for extensibility, as we believe this to be better for contributions to our various plugins.

## New Terminal UI

With Nx being a CLI tool, the terminal is our main interface with developers, so we want to make it as intuitive and efficient as possible. To this end, we will be exploring improvements to the terminal experience, especially when it comes to `nx run-many` and `nx affected`. This work is also a part of our migration of Nx core to Rust. There are Rust crates that we think will enable us to provide a more powerful and interactive terminal experience.

## Better and Faster TypeScript Monorepos

We will continue to invest time to make Nx the best tool for large TypeScript monorepos.

In Nx 20, we collaborated with the TypeScript Team and updated our [TypeScript monorepo setup](https://nx.dev/blog/announcing-nx-20#typescript-project-references-for-monorepos) to align with best practices in modern monorepos. We now use TypeScript project references to provide faster builds and typechecks. It also allows editors, such as VSCode, to better understand the monorepo and pick up code references more accurately. Additionally, we also enable [npm](https://docs.npmjs.com/cli/using-npm/workspaces)/[pnpm](https://pnpm.io/workspaces)/[yarn](https://yarnpkg.com/features/workspaces)/[bun](https://bun.sh/docs/install/workspaces) workspaces for new Nx workspaces. This feature is used by many monorepos, and provides a standard way to link and manage versions of packages.

We've already started extending this new setup to other official Nx plugins, such as `@nx/react` and `@nx/angular`, and should complete this work before 2025. The next step is to iterate on this new setup based on real-world usage and user feedback.

## Node.js Backend Deployments

Previously, we provided basic features to support Node.js backend development within Nx monorepos. We want to improve on this experience to offer more built-in capabilities, especially when it comes to deployments.

We will be looking at improved Docker support, including lockfile pruning for a subset of the workspace lockfile to build and run backend applications.

We also want to explore deployments of backend applications via `nx release`. Currently, the release feature is primarily used for publishing libraries/packages/crates to a relevant registry, but its API is generic and can be used for application deployments as well.

We will also look into other Node.js backend features, such as HMR for server code.

## Smarter Monorepos with AI

Generative AI can be a huge time-saver, especially in large code bases. We will explore using AI to write migrations and generators. These features can save hours, if not days, of work for large monorepos, especially within enterprise organizations.

## Expanded Java Support

We released `@nx/gradle` in Nx 19, which allows Nx to run Gradle tasks. We saw good results with using [task splitting](https://nx.dev/ci/features/split-e2e-tasks) to distribute tests across multiple agents in CI.

The next step is to iterate on our Gradle support based on user feedback. We want more Gradle users to use Nx to speed up their CI. We are also looking into Maven support in order to support even more Java projects.

## Module Federation 2.0

We want to make Nx the defacto tool for managing projects using Module Federation, which has been invaluable to many enterprise teams that we've closely collaborated with.

The Nx team has worked closely with the [Module Federation](https://module-federation.io/) team to integrate new features from Module Federation 2.0, such as Federation Runtime to manage dynamic remotes. This means that the older [dynamic remote feature](https://nx.dev/recipes/angular/dynamic-module-federation-with-angular) is no longer necessary and will be deprecated in 2025.

From 2023 to 2024, we've seen a lot of success with the adoption of our module federation tooling within enterprise organizations, and we will continue to iterate on this based on both user feedback and feedback from the Module Federation team.

**Comment by jame-earnin** [comment]
What about the nx release for **multiple apps** with integrated-base monorepo (Single package.json)?

**Reply by jaysoo** [reply]
There will be features and enhancements outside of what's on this roadmap. This roadmap provides key focus areas. If you have any feature requests, we recommend starting [a discussion](https://github.com/nrwl/nx/discussions/new?category=feature-requests).

**Reply by joshdales** [reply]
There is a discussion for this https://github.com/nrwl/nx/discussions/23389

**Comment by EelcoLos** [comment]
I recall @vsavkin  saying MonoRepo World that NX itself should be language agnostic, ie: any type of application should be add-able to NX. The latest announcement of Gradle first party support would be part of this, I guess.
However, I see no roadmap for first party support in 2025. 
For example, I'd suggest integrating `dotnet` with first party support.
What is the vision there?

**Reply by JamesHenry** [reply]
@EelcoLos Nx is already language agnostic and can be used with any stack. With our agnostic core moving to be fully written in rust, the requirement for Node is completely removed, which further strengthens this. 

The experience of using different stacks becomes much better if dedicated plugins are written for them, so that the agnostic core can be informed on how to discover projects, build the dependency graph etc.

It does not, however, mean that the Nx Team themselves will directly provide plugins for every stack. As you said, Gradle is already provided by the Nx Team, but there are other great community plugins out there already for other languages/ecosystems:

**Golang**: https://github.com/nx-go/nx-go
**Rust**: https://github.com/Cammisuli/monodon/blob/main/packages/rust/README.md (Nx Employee in their spare time)
**Dotnet**: https://github.com/nx-dotnet/nx-dotnet (Nx Employee in their spare time)
**Python**: https://github.com/lucasvieirasilva/nx-plugins/tree/main/packages/nx-python

You can discover these and many more in the plugin registry: https://nx.dev/plugin-registry

If there is super strong demand from our enterprise customers for a particular toolchain then we will of course prioritize that, and that is how our Gradle plugin came to be in the first place.

Hope that helps!

**Reply by pumano** [reply]
Would love if nx move dependency graph logic to rust too, it's open possibility to you guys port your nx/enforce-module-boundaries to oxlint / biome too

**Comment by lppedd** [comment]
> And I also want to reiterate that we will continue to use TypeScript for extensibility, as we believe this to be better for contributions to our various plugins.

Just want to be sure about current extensibility. We have more than 20 custom plugins, and many more to come, should I keep going or will the plugin/executor/generator API change with the migration to Rust? I don't really like the idea of having to rewrite thousands of lines of code. Thanks! 😄

**Reply by JamesHenry** [reply]
It is not our intention to change those APIs as part of this. If it were necessary or meaningfully beneficial we would provide a newly versioned implementation and deprecate the old one and maintain both for a time (we already did this successfully with CreateNodesV2)

**Comment by kkuriata** [comment]
Any plans of introducing Bun app/test runtimes and bundler?

**Comment by kkuriata** [comment]
What about further SWC support? It would be nice to have a fast compiler in huge apps. It would save a lot of CI (and developers') time.

**Reply by pumano** [reply]
Vite add their native oxc compiler in 2025 and you got benefits. Just wait. No need to check out swc which is 3x slower than oxc

**Reply by kkuriata** [reply]
Oh, that's nice! Thanks for answer!

**Comment by ceremonious** [comment]
Are there any near-term plans to enable Typescript Project References for existing NX repos?

**Reply by JamesHenry** [reply]
We are currently in the process of migrating repos of different shapes and sizes that we control or directly influence. It may not be possible to provide a robust automated migration for this change, but we will be able to produce some helpful guides on how to do it yourself once we have completed those efforts.

**Reply by ceremonious** [reply]
Thank you, yes any guides you're able to provide would be greatly appreciated. Our Typescript performance has degraded significantly in our mono-repo and we're looking for anything that will help

**Reply by techfg** [reply]
> We are currently in the process of migrating repos of different shapes and sizes that we control or directly influence. It may not be possible to provide a robust automated migration for this change, but we will be able to produce some helpful guides on how to do it yourself once we have completed those efforts.

@JamesHenry - Understood regarding the complexity of automating this and glad to hear that there are plans for providing guides once the efforts are completed.  In the meantime, is it possible to share some links on projects that have gone through this thus far and/or in the process of going through it?  Having a reference, even if just commits/source would be helpful until more information is available.  For example, I saw that [nx-console](https://github.com/nrwl/nx-console/commit/74453086f6abead5325270cf73abac2bbc42ec12) appears to have been migrated but if you are able to provide any additional pointers that would be helpful.  Thank you!

**Reply by JamesHenry** [reply]
Whilst going through these manual migrations, @jaysoo has managed to create a generator that does 80-90% of the work automatically, more info will follow on this soon

**Reply by sergio-milu** [reply]
any update in that @jaysoo generator? would be great to have some automation here, even if it is not a 100%, 80% would be lovely

**Comment by christianulson** [comment]
How does Nx affected work with Git submodules? Here we use them quite a bit, since each sub-project’s modules are in different repositories, which makes organization and security easier.

**Reply by byjokese** [reply]
I have some git submodules, and I like the approach even though they give many headaches with team operations and complex branching strategies. Haven't used them yet with NX and frontend, have some on backend projects.

**Comment by moatorres** [comment]
I'd be happy to contribute to the app deployment initiative, I've been generating/hydrating `Dockerfile`s and Kubernetes YAML manifests for years with `nx` generators. It is incredibly powerful. 🚀

**Comment by khludenevav** [comment]
To simplify migration to nx and increase adoption, can team add an option to ignore some `project.json` files (for now all settings of projects inside project.json)? Because my repo has these files. I can't rename them.
Details: https://github.com/nrwl/nx/discussions/27299

**Comment by rathpc** [comment]
Do you have a general estimate for when you think v21 will drop?

**Reply by JamesHenry** [reply]
> Major Nx versions are released as the latest every six months, typically around April and October

https://nx.dev/reference/releases

**Comment by byjokese** [comment]
Cant wait for Module Federation 2.0, and integration with the latest angular builder: esbuild and vite. Any matter on this subject is gold.

Also thanks for all the work and good luck with the rust migration, seem really promising for the future ❤️.

**Comment by igortas** [comment]
Explore more time to make ```nx``` more stable, because in compare to turborepo, it's terrible how bad experience nx provide. So all other parts can be halted. Try to write test for as many possible scenarios, because there is no single day where nx is not failing. U can't even start basic commands like build, test or lint, that's how bad experience with nx is. I can open thousand tickets, what everything I found.

**Reply by moatorres** [reply]
Turborepo is awesome. 🚀 What specific failures are you referring to? I've just started a new project with Nx without issues, maybe I can help you.

**Reply by igortas** [reply]
@moatorres Nearly everything. I'm just doing some demo projects, and everytime, something else is breaking. I never seen Nx stable for more than 10 minutes working. If it's cache issue or something else. Turborepo works everything out of the box. I don't build anything wow, just 2-3 microservices and single frontend project. It's been long since, I've play around, I don't think will something change now. Or I'm expecting too much custom things from Nx, that it's not capable of. I will try to build someday the same project in nx and in turborepo, for better repro, and my point that Nx is way unstable. I've opened some tickets, and all were just closed.

**Reply by moatorres** [reply]
I’m glad Turborepo works well for you, it’s a solid tool. At the same time, I don’t think it’s fair to call Nx inherently unstable. There are many teams and large companies using it successfully in production every day, myself included. For us it has been extremely reliable and a huge productivity boost.

That doesn’t mean your experience isn’t real, just that it may be tied to configuration, workflow, or expectations rather than Nx being fundamentally broken. It can be a bit more opinionated and feature-rich than Turborepo, so there’s naturally a bit more to learn.

If you ever decide to revisit your setup or share a reproducible example, I’m sure the community would be happy to help figure out what went wrong. But saying Nx is “unstable” as a whole doesn’t really reflect the broader reality, IMHO.

**Comment by ianzone** [comment]
Would you support fractions for `parallel`? https://github.com/nrwl/nx/discussions/30238

**Comment by HaasStefan** [comment]
Is there already a roadmap for 2026?

**Reply by byjokese** [reply]
Would be interesting to now the progress of 2025, conclusions and ideas for 2026

**Reply by zargham-k** [reply]
+1, would be really interesting to find out where nx is headed!
